source/UF/Truncations.lagda

```diff
@@ -0,0 +1,428 @@
+            (∣ x ∣[ n ])                         ∎
+      where
+       I = ap b (∥∥ₙ-rec-comp (∥∥ₙ-h-level n) (λ x → ∣ ∣ x ∣[ succ n ] ∣[ n ]) x)
+       II = ∥∥ₙ-rec-comp (∥∥ₙ-h-level n) canonical-pred-map (∣ x ∣[ succ n ])
+
+\end{code}
+
+We now define an equivalence that characterizes the truncated identity type
+under the assumption of univalence.
+
+\begin{code}
+
+ canonical-identity-trunc-map : {𝓤 : Universe} {X : 𝓤 ̇} {x x' : X} {n : ℕ}
+                              → ∥ x ＝ x' ∥[ n ]
+                              → ∣ x ∣[ succ n ] ＝ ∣ x' ∣[ succ n ]
+ canonical-identity-trunc-map {𝓤} {X} {x} {x'} {n} =
+  ∥∥ₙ-rec (∥∥ₙ-h-level (succ n) ∣ x ∣[ succ n ] ∣ x' ∣[ succ n ])
+          (ap (λ x → ∣ x ∣[ (succ n) ]))
+
+ module _ {𝓤 : Universe} {X : 𝓤 ̇} {n : ℕ}
+          (ua : Univalence) (x : X) 
+           where
+
+  trunc-id-family : ∥ X ∥[ succ n ] → ℍ n 𝓤
+  trunc-id-family = ∥∥ₙ-rec {𝓤} {𝓤 ⁺} {X} {ℍ n 𝓤} {succ n}
+                            (ℍ-is-of-next-hlevel n 𝓤 (ua 𝓤))
+                            (λ x' → (∥ x ＝ x' ∥[ n ] , (∥∥ₙ-h-level n)))
+
+  trunc-id-family-type : ∥ X ∥[ succ n ] → 𝓤 ̇
+  trunc-id-family-type = pr₁ ∘ trunc-id-family
+
+  trunc-id-family-level : (v : ∥ X ∥[ succ n ])
+                        → (trunc-id-family-type v) is-of-hlevel n
+  trunc-id-family-level = pr₂ ∘ trunc-id-family
+
+  trunc-id-family-computes : (x' : X)
+                           → trunc-id-family-type ∣ x' ∣[ succ n ]
+                             ＝ ∥ x ＝ x' ∥[ n ]
+  trunc-id-family-computes x' =
+    ap pr₁ (∥∥ₙ-rec-comp (ℍ-is-of-next-hlevel n 𝓤 (ua 𝓤))
+                         (λ x' → (∥ x ＝ x' ∥[ n ] , (∥∥ₙ-h-level n)))
+                         x')
+
+  trunc-id-forward-map : (x' : X)
+                       → trunc-id-family-type ∣ x' ∣[ succ n ]
+                       → ∥ x ＝ x' ∥[ n ]
+  trunc-id-forward-map x' = transport id (trunc-id-family-computes x')
+
+  trunc-id-backward-map : (x' : X)
+                        → ∥ x ＝ x' ∥[ n ]
+                        → trunc-id-family-type ∣ x' ∣[ succ n ]
+  trunc-id-backward-map x' = transport⁻¹ id (trunc-id-family-computes x')
+
+  trunc-id-back-is-retraction : (x' : X)
+                              → trunc-id-backward-map x' ∘ trunc-id-forward-map x'
+                               ∼ id
+  trunc-id-back-is-retraction x' q =
+   forth-and-back-transport (trunc-id-family-computes x')
+
+  refl-trunc-id-family : trunc-id-family-type ∣ x ∣[ succ n ]
+  refl-trunc-id-family = trunc-id-backward-map x ∣ refl ∣[ n ]
+
+  identity-on-trunc-to-family : (v : ∥ X ∥[ succ n ])
+                              → (∣ x ∣[ succ n ] ＝ v)
+                              → trunc-id-family-type v
+  identity-on-trunc-to-family .(∣ x ∣[ succ n ]) refl = refl-trunc-id-family
+
+  trunc-id-family-is-identity-system : is-contr (Σ (trunc-id-family-type))
+  trunc-id-family-is-identity-system =
+   ((∣ x ∣[ succ n ] , refl-trunc-id-family) , center-Q)
+   where
+    I : (x' : X) (p : x ＝ x')
```

Comment:
Do these steps implement the proof I sketched on Discord? I would've thought that it should be simpler than the below. Did I miss something?

Reply:
The proof you outlined was a string of equivalences but as you noted you really only needed functions in one of the directions. What I implemented here is that string of functions.

Reply:
Right, I got that, but somehow the code seems too complicated, although I may be underestimating the conversion from paper to Agda. I'll probably take another stab at it in Round 2.

Reply:
It wouldn't be the first time that I've overcomplicated something. You likely anticipate that I have to curry and apply truncation induction for two variables, you may be overlooking that another issue is that my trunc-id family only computes up to homotopy so I have to transport along that homotopy (your proof was able to gloss over this fact).

Reply:
Got ya. I also had a go and I see where the complications came in. I'll fix up the indentation myself after merging this to try and make it a bit more readable, but we can consider this resolved.

Reply:
Okay